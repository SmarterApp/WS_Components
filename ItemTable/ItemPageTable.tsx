import * as React from 'react'
import * as ItemTableHeader from './ItemTableHeader'
import * as ItemTable from './ItemTable'
import * as ItemCardViewModel from '../Models/ItemCardViewModel'
import * as GradeLevels from "../Models/GradeLevels";
import { ItemsSearchViewModel } from "../ScoreGuide/ItemSearchContainer";
import * as ItemModels from '../Models/ItemModels';
import * as ItemSearchDropdown from '../DropDown/ItemSearchDropdown';


export interface Props {
    onRowSelection: (item: {itemKey: number; bankKey: number}) => void;
    itemCards?: ItemCardViewModel.ItemCardViewModel[]; 
}

export interface State {
    sorts: ItemTableHeader.HeaderSort[]; 
    selectedRow?: ItemCardViewModel.ItemCardViewModel; 
    
}

export class ItemPageTable extends React.Component<Props, State>{
    private headerColumns = ItemTableHeader.headerColumns;
    private dataTableRef: HTMLTableElement;

    constructor(props: Props) {
        super(props); 
        this.state = {
            sorts: []
        }
    }

    onClickHeader = (col: ItemTableHeader.SortColumn) => {
        const newSorts = (this.state.sorts || []).slice();
        const headIdx = newSorts.findIndex(hs => hs.col.header === col.header);
        if (headIdx !== -1) {
            const newSort = Object.assign({}, newSorts[headIdx]);
            if (newSort.direction == ItemTableHeader.SortDirection.Ascending) {
                newSort.direction = ItemTableHeader.SortDirection.Descending;
            }
            else if (newSort.direction == ItemTableHeader.SortDirection.Descending) {
                newSort.direction = ItemTableHeader.SortDirection.NoSort;
            }
            else {
                newSort.direction = ItemTableHeader.SortDirection.Ascending;
            }

            newSorts[headIdx] = newSort;
        } else {
            const newSort: ItemTableHeader.HeaderSort = {
                col: col,
                direction: ItemTableHeader.SortDirection.Ascending,
                resetSortCount: 0
            };
            newSorts.push(newSort);
        }
        this.setState({ sorts: newSorts });
    }

    onSelectItem = (item: ItemCardViewModel.ItemCardViewModel) => {
        this.setState({
            selectedRow: item
        });
        const card = {itemKey: item.itemKey, bankKey: item.bankKey}

        this.props.onRowSelection(card);
     
    };

    invokeMultiSort(lhs: ItemCardViewModel.ItemCardViewModel, rhs: ItemCardViewModel.ItemCardViewModel): number {
        const sorts = this.state.sorts || [];
        for (const sort of sorts) {
            const diff = sort.col.compare(lhs, rhs) * sort.direction;
            if (diff !== 0) {
                return diff;
            }
        }
        return 0;
    }
    getTableData = (): ItemCardViewModel.ItemCardViewModel[] | undefined => {
        let data = this.props.itemCards; 
        if(data != undefined){
            data = this.state.sorts && this.state.sorts.length !== 0
                ? data.sort((lhs, rhs) => this.invokeMultiSort(lhs, rhs))
                : data;
        }
        return data; 
    }
    
    renderTableHeader() {
        return (
            <ItemTableHeader.HeaderTable
                sorts={this.state.sorts}
                onHeaderClick={this.onClickHeader}
                columns={this.headerColumns} />
        );
    }

    renderTable() {
        const itemCards = this.getTableData(); //this returns undefined but in the method it has data. that's w
        if (itemCards != undefined) {
            //if no items are returned we want to return a friendly message
            if (itemCards.length === 0) {
                return <span className="placeholder-text" role="alert">No results found for the given search terms.</span>
            } else {
                return (
                    <ItemTable.DataTable
                        mapRows={itemCards}
                        rowOnClick={this.onSelectItem}
                        sort={this.state.sorts}
                        tableRef={ref => this.dataTableRef = ref}
                        columns={this.headerColumns}
                        selectedRow={this.state.selectedRow} />
                );
            }
        }
        return <span className="placeholder-text" role="alert">No results found for the given search terms.</span>
    }

    render() {
        return (
            <div>
                {this.renderTableHeader()};
                {this.renderTable()};
            </div>
        );
    }
}
