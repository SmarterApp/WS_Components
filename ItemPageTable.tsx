import * as React from 'react'
import * as ItemTableHeader from './ItemTableHeader'
import * as ItemTable from './ItemTable'
import * as ApiModels from './ApiModels'
import * as ItemCardViewModel from './ItemCardViewModel'
import * as AboutItem from './AboutItem'
import * as GradeLevels from "./GradeLevels";
import { get } from "./ApiModels";
import { ItemsSearchViewModel } from "./ItemPageSearch";
import * as ItemModels from './ItemModels';
import * as ItemSearchDropdown from './ItemSearchDropdown';



const SearchClient = (params: ItemSearchDropdown.SearchAPIParams) => get<ItemCardViewModel.ItemCardViewModel[]>("http://is-score.cass.oregonstate.edu/ScoringGuide/Search", params);

export interface Props {
    onRowSelection: (item: {itemKey: number; bankKey: number}) => void;
    scoringGuideViewModel?: ItemsSearchViewModel; //what do we need this for in this file?
}

export interface State {
    searchParams: ItemModels.ScoreSearchParams; //nah dawg 
    sorts: ItemTableHeader.HeaderSort[]; //should this be state?
    itemSearchResult: ApiModels.Resource<ItemCardViewModel.ItemCardViewModel[]>; //I've been asking myself that.  I think there are a few things messed up about the design
    selectedRow?: ItemCardViewModel.ItemCardViewModel; //I did not like this implementation but whatever
    
}
export class ItemPageTable extends React.Component<Props, State>{
    private headerColumns = ItemTableHeader.headerColumns;
    private dataTableRef: HTMLTableElement;

    constructor() {
        super(); 
        this.state = {
            searchParams: {
                gradeLevels: GradeLevels.GradeLevels.NA,
                subjects: [],
                techType: []
            },
            itemSearchResult: { kind: "none" },
            sorts: [],
        }

        //court, add the search method to get items. 

    }
   

    onClickHeader = (col: ItemTableHeader.SortColumn) => {
        const newSorts = (this.state.sorts || []).slice();
        const headIdx = newSorts.findIndex(hs => hs.col.header === col.header);
        if (headIdx !== -1) {
            const newSort = Object.assign({}, newSorts[headIdx]);
            if (newSort.direction == ItemTableHeader.SortDirection.Ascending) {
                newSort.direction = ItemTableHeader.SortDirection.Descending;
            }
            else if (newSort.direction == ItemTableHeader.SortDirection.Descending) {
                newSort.direction = ItemTableHeader.SortDirection.NoSort;
            }
            else {
                newSort.direction = ItemTableHeader.SortDirection.Ascending;
            }

            newSorts[headIdx] = newSort;
        } else {
            const newSort: ItemTableHeader.HeaderSort = {
                col: col,
                direction: ItemTableHeader.SortDirection.Ascending,
                resetSortCount: 0
            };
            newSorts.push(newSort);
        }
        this.setState({ sorts: newSorts });
    }

    onSelectItem = (item: ItemCardViewModel.ItemCardViewModel) => {
        this.setState({
            selectedRow: item
        });
        const card = {itemKey: item.itemKey, bankKey: item.bankKey}

        this.props.onRowSelection(card);
     
    };

  

    invokeMultiSort(lhs: ItemCardViewModel.ItemCardViewModel, rhs: ItemCardViewModel.ItemCardViewModel): number {
        const sorts = this.state.sorts || [];
        for (const sort of sorts) {
            const diff = sort.col.compare(lhs, rhs) * sort.direction;
            if (diff !== 0) {
                return diff;
            }
        }
        return 0;
    }

    getTableData(data: ItemCardViewModel.ItemCardViewModel[]): ItemCardViewModel.ItemCardViewModel[] {
        const sortedData = this.state.sorts && this.state.sorts.length !== 0
            ? data.sort((lhs, rhs) => this.invokeMultiSort(lhs, rhs))
            : data;

        return sortedData;
    }

    renderTableHeader() {
        return (
            <ItemTableHeader.HeaderTable
                sorts={this.state.sorts}
                onHeaderClick={this.onClickHeader}
                columns={this.headerColumns} />
        );
    }

    renderTable() {
        const result = this.state.itemSearchResult;
        let resultElement: JSX.Element[] | JSX.Element | undefined;
        if (result.kind == "success" || result.kind == "reloading") {
            if (result.content == null || result.content.length === 0) {
                resultElement = <span className="placeholder-text" role="alert">No results found for the given search terms.</span>
            } else {
                return (
                    <ItemTable.DataTable
                        mapRows={this.getTableData(result.content)}
                        rowOnClick={this.onSelectItem}
                        sort={this.state.sorts}
                        tableRef={ref => this.dataTableRef = ref}
                        columns={this.headerColumns}
                        selectedRow={this.state.selectedRow} />
                );
            }
        } else {
            resultElement = <div className="placeholder-text" role="alert">An error occurred. Please try again later.</div>;
        }
        return resultElement;
    }

    render() {
        return (
            <div>
                {this.renderTableHeader()};
                {this.renderTable()};
            </div>
        );
    }
}
