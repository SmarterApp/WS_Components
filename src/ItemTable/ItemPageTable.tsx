import * as React from 'react'
import * as ItemTableHeader from './ItemTableHeader'
import * as ItemTable from './ItemTable'
import * as ItemCardViewModel from '../Models/ItemCardViewModel'
import * as GradeLevels from "../Models/GradeLevels";
import { ItemsSearchViewModel } from "../ScoreGuide/ItemSearchContainer";
import * as ItemModels from '../Models/ItemModels';
import * as AboutItemVM from '../Models/AboutItemVM';
import * as ApiModels from '../Models/ApiModels';

export interface Props {
    onRowSelection: (item: { itemKey: number; bankKey: number }, reset: boolean) => void;
    itemCards?: ItemCardViewModel.ItemCardViewModel[];
    item: ApiModels.Resource<AboutItemVM.AboutThisItem>;
}

export interface State {
    sorts: ItemTableHeader.HeaderSort[];
    selectedRow?: ItemCardViewModel.ItemCardViewModel | null;
}

export class ItemPageTable extends React.Component<Props, State>{
    private headerColumns = ItemTableHeader.headerColumns;
    private dataTableRef: HTMLTableElement;

    constructor(props: Props) {
        super(props);
        this.state = {
            sorts: [],
            selectedRow: null
        }
    }

    onClickHeader = (col: ItemTableHeader.SortColumn) => {
        const newSorts = (this.state.sorts || []).slice();
        const headIdx = newSorts.findIndex(hs => hs.col.header === col.header);
        if (headIdx !== -1) {
            const newSort = Object.assign({}, newSorts[headIdx]);
            if (newSort.direction == ItemTableHeader.SortDirection.Ascending) {
                newSort.direction = ItemTableHeader.SortDirection.Descending;
            }
            else if (newSort.direction == ItemTableHeader.SortDirection.Descending) {
                newSort.direction = ItemTableHeader.SortDirection.NoSort;
            }
            else {
                newSort.direction = ItemTableHeader.SortDirection.Ascending;
            }
            newSorts[headIdx] = newSort;
        } else {
            const newSort: ItemTableHeader.HeaderSort = {
                col: col,
                direction: ItemTableHeader.SortDirection.Ascending,
                resetSortCount: 0
            };
            newSorts.push(newSort);
        }
        this.setState({ sorts: newSorts });
    }

    onSelectItem = (item: ItemCardViewModel.ItemCardViewModel) => {
        const card = { itemKey: item.itemKey, bankKey: item.bankKey }
        if(item === this.state.selectedRow){
            this.props.onRowSelection(card, true)
            this.setState({selectedRow: null})
        }else{
            this.props.onRowSelection(card, false)
            this.setState({selectedRow: item})
        }
    };

    invokeMultiSort(lhs: ItemCardViewModel.ItemCardViewModel, rhs: ItemCardViewModel.ItemCardViewModel): number {
        const sorts = this.state.sorts || [];
        for (const sort of sorts) {
            const diff = sort.col.compare(lhs, rhs) * sort.direction;
            if (diff !== 0) {
                return diff;
            }
        }
        return 0;
    }
    
    getTableData = (): ItemCardViewModel.ItemCardViewModel[] | undefined => {
        let data = this.props.itemCards;
        if (data != undefined) {
            data = this.state.sorts && this.state.sorts.length !== 0
                ? data.sort((lhs, rhs) => this.invokeMultiSort(lhs, rhs))
                : data;
        }
        return data;
    }

    renderTableHeader() {
        return (
            <ItemTableHeader.HeaderTable
                sorts={this.state.sorts}
                onHeaderClick={this.onClickHeader}
                columns={this.headerColumns} />
        );
    }

    renderTable() {
        const itemCards = this.getTableData(); //this returns undefined but in the method it has data. that's w
        let content = (<span className="placeholder-text" role="alert">No results found for the given search terms.</span>);
        if (itemCards != undefined) {
            //if no items are returned we want to return a friendly message
            if (itemCards.length !== 0) {
                content = (
                    <ItemTable.DataTable
                        mapRows={itemCards}
                        rowOnClick={this.onSelectItem}
                        sort={this.state.sorts}
                        tableRef={ref => this.dataTableRef = ref}
                        columns={this.headerColumns}
                        selectedRow={this.state.selectedRow}
                        item={this.props.item}
                    />
                );
            }
        }
        return content;
    }

    render() {
        return (
            <div>
                <table className="item-table mapcomponent-table">
                    {this.renderTableHeader()}
                    {this.renderTable()}
                </table>
            </div>
        );
    }
}
