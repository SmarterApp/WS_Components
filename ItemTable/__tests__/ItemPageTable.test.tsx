import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { ItemPageTable } from '../ItemPageTable';
import * as TestUtils from 'react-dom/test-utils';
import { shallow, mount } from 'enzyme';

describe("ItemPageTable", () => {
    const props = {
        onRowSelection: jest.fn(),
        itemCards: [] 
    }

    it("matches snapshot", () => {
        expect(shallow(<ItemPageTable {...props}/>)).toMatchSnapshot();;
    })
})