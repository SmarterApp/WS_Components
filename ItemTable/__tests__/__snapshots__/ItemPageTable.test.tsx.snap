// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemPageTable matches snapshot 1`] = `
ShallowWrapper {
  "complexSelector": ComplexSelector {
    "buildPredicate": [Function],
    "childrenOfNode": [Function],
    "findWhereUnwrapped": [Function],
  },
  "length": 1,
  "node": <div>
    <HeaderTable
        columns={
            Array [
                Object {
                  "accessor": [Function],
                  "className": "item",
                  "compare": [Function],
                  "header": "Item",
                },
                Object {
                  "accessor": [Function],
                  "className": "claimAndTarget",
                  "compare": [Function],
                  "header": "Claim/Target",
                },
                Object {
                  "accessor": [Function],
                  "className": "subject",
                  "compare": [Function],
                  "header": "Subject",
                },
                Object {
                  "accessor": [Function],
                  "className": "grade",
                  "compare": [Function],
                  "header": "Grade",
                },
                Object {
                  "accessor": [Function],
                  "className": "interactionType",
                  "compare": [Function],
                  "header": "Item Type",
                },
              ]
        }
        onHeaderClick={[Function]}
        sorts={Array []}
    />
    <span
        className="placeholder-text"
        role="alert"
    >
        No results found for the given search terms.
    </span>
</div>,
  "nodes": Array [
    <div>
      <HeaderTable
            columns={
                  Array [
                        Object {
                          "accessor": [Function],
                          "className": "item",
                          "compare": [Function],
                          "header": "Item",
                        },
                        Object {
                          "accessor": [Function],
                          "className": "claimAndTarget",
                          "compare": [Function],
                          "header": "Claim/Target",
                        },
                        Object {
                          "accessor": [Function],
                          "className": "subject",
                          "compare": [Function],
                          "header": "Subject",
                        },
                        Object {
                          "accessor": [Function],
                          "className": "grade",
                          "compare": [Function],
                          "header": "Grade",
                        },
                        Object {
                          "accessor": [Function],
                          "className": "interactionType",
                          "compare": [Function],
                          "header": "Item Type",
                        },
                      ]
            }
            onHeaderClick={[Function]}
            sorts={Array []}
      />
      <span
            className="placeholder-text"
            role="alert"
      >
            No results found for the given search terms.
      </span>
</div>,
  ],
  "options": Object {},
  "renderer": ReactShallowRenderer {
    "_instance": ShallowComponentWrapper {
      "_calledComponentWillUnmount": false,
      "_compositeType": 0,
      "_context": Object {},
      "_currentElement": <ItemPageTable
        itemCards={Array []}
        onRowSelection={[Function]}
/>,
      "_debugID": 1,
      "_hostContainerInfo": null,
      "_hostParent": null,
      "_instance": ItemPageTable {
        "_reactInternalInstance": [Circular],
        "context": Object {},
        "getTableData": [Function],
        "headerColumns": Array [
          Object {
            "accessor": [Function],
            "className": "item",
            "compare": [Function],
            "header": "Item",
          },
          Object {
            "accessor": [Function],
            "className": "claimAndTarget",
            "compare": [Function],
            "header": "Claim/Target",
          },
          Object {
            "accessor": [Function],
            "className": "subject",
            "compare": [Function],
            "header": "Subject",
          },
          Object {
            "accessor": [Function],
            "className": "grade",
            "compare": [Function],
            "header": "Grade",
          },
          Object {
            "accessor": [Function],
            "className": "interactionType",
            "compare": [Function],
            "header": "Item Type",
          },
        ],
        "onClickHeader": [Function],
        "onSelectItem": [Function],
        "props": Object {
          "itemCards": Array [],
          "onRowSelection": [Function],
        },
        "refs": Object {},
        "state": Object {
          "sorts": Array [],
        },
        "updater": Object {
          "enqueueCallback": [Function],
          "enqueueCallbackInternal": [Function],
          "enqueueElementInternal": [Function],
          "enqueueForceUpdate": [Function],
          "enqueueReplaceState": [Function],
          "enqueueSetState": [Function],
          "isMounted": [Function],
          "validateCallback": [Function],
        },
      },
      "_mountOrder": 1,
      "_pendingCallbacks": null,
      "_pendingElement": null,
      "_pendingForceUpdate": false,
      "_pendingReplaceState": false,
      "_pendingStateQueue": null,
      "_renderedComponent": NoopInternalComponent {
        "_currentElement": <div>
          <HeaderTable
                    columns={
                              Array [
                                        Object {
                                          "accessor": [Function],
                                          "className": "item",
                                          "compare": [Function],
                                          "header": "Item",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "claimAndTarget",
                                          "compare": [Function],
                                          "header": "Claim/Target",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "subject",
                                          "compare": [Function],
                                          "header": "Subject",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "grade",
                                          "compare": [Function],
                                          "header": "Grade",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "interactionType",
                                          "compare": [Function],
                                          "header": "Item Type",
                                        },
                                      ]
                    }
                    onHeaderClick={[Function]}
                    sorts={Array []}
          />
          <span
                    className="placeholder-text"
                    role="alert"
          >
                    No results found for the given search terms.
          </span>
</div>,
        "_debugID": 2,
        "_renderedOutput": <div>
          <HeaderTable
                    columns={
                              Array [
                                        Object {
                                          "accessor": [Function],
                                          "className": "item",
                                          "compare": [Function],
                                          "header": "Item",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "claimAndTarget",
                                          "compare": [Function],
                                          "header": "Claim/Target",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "subject",
                                          "compare": [Function],
                                          "header": "Subject",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "grade",
                                          "compare": [Function],
                                          "header": "Grade",
                                        },
                                        Object {
                                          "accessor": [Function],
                                          "className": "interactionType",
                                          "compare": [Function],
                                          "header": "Item Type",
                                        },
                                      ]
                    }
                    onHeaderClick={[Function]}
                    sorts={Array []}
          />
          <span
                    className="placeholder-text"
                    role="alert"
          >
                    No results found for the given search terms.
          </span>
</div>,
      },
      "_renderedNodeType": 0,
      "_rootNodeID": 0,
      "_topLevelWrapper": null,
      "_updateBatchNumber": null,
      "_warnedAboutRefsInRender": false,
    },
    "getRenderOutput": [Function],
    "render": [Function],
  },
  "root": [Circular],
  "unrendered": <ItemPageTable
    itemCards={Array []}
    onRowSelection={[Function]}
/>,
}
`;
